feat(event): add parent hash in the block event

clarkchen opened the pull request:
**What does this PR do?**
Adds parent hash information to BlockLogTrigger to include this data in the block streaming output.


**Why are these changes required?**
This enhancement helps users easily identify forked blockchain scenarios by providing the parent hash information in the block logs, which is crucial for blockchain analysis and troubleshooting.


**This PR has been tested by:**
- Manual Testing

**Follow up**
NO

**Extra details**
The implementation adds a parentHash field to BlockLogTrigger and populates it in BlockLogTriggerCapsule by retrieving the parent hash from the block data.

clarkchen committed:
feat: add parent hash in the block event

clarkchen commented:
@waynercheung good point.

I have checked the the code, the new parent hash will also be compatible with the event plugin, because java_tron passes objects to the plugin in string format.



##  EventPluginLoader.java 
* **parent hash** will be appened to the return **toJsonString(trigger)**

https://github.com/tronprotocol/java-tron/blob/2cc57bd1535d157d2a9f27c3fb77dd06f872c726/framework/src/main/java/org/tron/common/logsfilter/EventPluginLoader.java#L464-L472


# Plugins
the Plugin Interface, e.g. KafkaEventListener.java, MongodbEventListener.java, will handle string smoothly and push to clients new message, which format is "[oldmsg], parentHash: xxxxxx"


https://github.com/tronprotocol/event-plugin/blob/97bea8cc200ee2a1b6abf61860b33782794fb192/plugins/kafkaplugin/src/main/java/org/tron/eventplugin/KafkaEventListener.java#L40-L45

clarkchen committed:
Merge branch 'tronprotocol:develop' into develop

clarkchen commented:
@waynercheung  

I've noticed the block event is used in the event-plugin and submit a new pr for you to review, thanks


https://github.com/tronprotocol/event-plugin/pull/47

clarkchen commented:
@CodeNinjaEvan  pls review this, thx~

0xbigapple commented:
A note on how `parentHash` field interacts with fork handling in the two event-subscription paths:

**V0 path**: In `Manager.pushBlock`, when a fork is detected, `switchFork` is called and `pushBlock` returns early — `postBlockTrigger` is not invoked. Inside `switchFork`, `eraseBlock` and `applyBlock(setSwitch(true))` also do not emit any `BlockLogTrigger`.

So a V0 subscriber sees: triggers for the losing-branch blocks were already delivered, the winning-branch replacement blocks do not produce triggers, and the next trigger is the first new block on the new main chain. With `parentHash` added, the subscriber can **detect** a fork by comparing against the previously recorded block hash; but the event stream neither revokes the already-delivered losing-branch blocks nor replays the winning-branch blocks, so reconciling derived data still requires the subscriber to query the node directly.

**V1 path**: `BlockEventLoad.load()` already walks back to the common ancestor via `BlockEvent.parentId`, emits orphaned blocks as `new Event(e, isRemove=true)`, and emits the winning branch in forward order as `new Event(e, false)`. Detect / revoke / replay are all handled internally, so `BlockLogTrigger.parentHash` is redundant for V1 consumers.

So, this PR lets V0 subscribers notice that a fork occurred, but does not provide the data needed to reconcile it; V1 already handles **Fock Problem** internally.

clarkchen commented:
@0xbigapple  Thanks for the detailed analysis. 

  This PR gives V0 subscribers the reliable way to detect same-height forks, at near-zero cost.

 **Why it matters**:
  - When a fork occurs at the same block height, block numbers stay sequential. Without parentHash, V0 subscribers have no signal that a fork happened — they silently operate on stale data.
  - parentHash doesn't solve full reconciliation, but it breaks the "silent corruption" scenario, letting subscribers know they need to re-sync.

  **Why the cost is justified:**
  - parentHash is an inherent block attribute — we're just exposing what BlockCapsule already has.
  - It aligns with industry conventions (e.g., Ethereum's newHeads subscription includes parentHash).
  - For V1, it's redundant but harmless — data completeness, not overhead.

0xbigapple commented on common/src/main/java/org/tron/common/logsfilter/trigger/BlockLogTrigger.java:
```diff
@@ -27,6 +27,11 @@ public class BlockLogTrigger extends Trigger {
   @Setter
   private List<String> transactionList = new ArrayList<>();
 
+
```
[NIT] Double blank line before the new field.

0xbigapple commented on common/src/main/java/org/tron/common/logsfilter/trigger/BlockLogTrigger.java:
```diff
@@ -45,6 +50,10 @@ public String toString() {
+        .append(", parentHash: ")
+        .append(parentHash)
+        .toString();
+
```
[NIT] Trailing blank line inside the toString() method body.

0xbigapple commented on common/src/main/java/org/tron/common/logsfilter/trigger/BlockLogTrigger.java:
```diff
@@ -27,6 +27,11 @@ public class BlockLogTrigger extends Trigger {
   private List<String> transactionList = new ArrayList<>();
 
+
+  @Getter
```
[SHOULD] Add an assertion that parentHash is populated by the capsule constructor.

0xbigapple commented:
This PR has many outdated CI checks. It’s recommended to merge the latest develop branch and reopen/retrigger the PR.

clarkchen committed:
Merge branch 'tronprotocol:develop' into develop

clarkchen committed:
style(event): address PR #6294 review nits and assert parentHash

- Remove double blank line before parentHash field in BlockLogTrigger
- Remove trailing blank line inside toString() body
- Add testParentHashPopulated to verify capsule constructor sets parentHash

clarkchen commented:
@0xbigapple  Thanks for the review! I've synced with the latest develop branch and addressed your comments in [42f0c6424](https://github.com/tronprotocol/java-tron/pull/6294/changes/42f0c6424c5d188d89cd0f6def4bb7937672b9a4)

0xbigapple commented:
@clarkchen The V0 revoke/replay gap described above has since been addressed by release v4.8.2 #6833 (merged).

On a fork switch, version-0 subscribers now get the full reconciliation data from the event stream itself:
- each erased losing-branch block re-emits its block/transaction triggers with removed=true (reOrgBlockTrigger);
- after the new branch is applied, its blocks' triggers are emitted forward (reApplyBlockEvents).

So V0 no longer requires querying the node to reconcile a reorg.

bladehan1 commented:
What are the future plans for this PR?

If there are no updates for an extended period, the PR will be closed.
